src/Components/Aspire.Microsoft.EntityFrameworkCore.SqlServer/AspireSqlServerEFCoreSqlClientExtensions.cs

```diff
@@ -59,15 +49,102 @@ public static class AspireSqlServerEFCoreSqlClientExtensions
+            });
+
+            configureDbContextOptions?.Invoke(dbContextOptionsBuilder);
         }
-        else
+    }
+
+    /// <summary>
+    /// Configures retries, health check, logging and telemetry for the <see cref="DbContext" />.
+    /// </summary>
+    /// <exception cref="ArgumentNullException">Thrown if mandatory <paramref name="builder"/> is null.</exception>
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="DbContext"/> is not registered in DI.</exception>
+    public static void EnrichSqlServerDbContext<[DynamicallyAccessedMembers(RequiredByEF)] TContext>(
+            this IHostApplicationBuilder builder,
+            Action<MicrosoftEntityFrameworkCoreSqlServerSettings>? configureSettings = null) where TContext : DbContext
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        var settings = GetDbContextSettings<TContext>(builder);
+
+        configureSettings?.Invoke(settings);
+
+        ConfigureRetry();
+
+        ConfigureInstrumentation<TContext>(builder, settings);
+
+        void ConfigureRetry()
         {
-            builder.Services.AddDbContext<TContext>(ConfigureDbContext);
+            if (!settings.Retry)
+            {
+                return;
+            }
+
+            // Resolving DbContext<TContextService> will resolve DbContextOptions<TContextImplementation>.
+            // We need to replace the DbContextOptions service descriptor to inject more logic. This won't be necessary once
+            // Aspire targets .NET 9 as EF will respect the calls to services.ConfigureDbContext<TContext>(). c.f. https://github.com/dotnet/efcore/pull/32518
+
+            var oldDbContextOptionsDescriptor = builder.Services.FirstOrDefault(sd => sd.ServiceType == typeof(DbContextOptions<TContext>));
+
+            if (oldDbContextOptionsDescriptor is null)
+            {
+                throw new InvalidOperationException($"DbContext<{typeof(TContext).Name}> was not registered");
+            }
+
+            builder.Services.Remove(oldDbContextOptionsDescriptor);
+
+            var dbContextOptionsDescriptor = new ServiceDescriptor(
+                oldDbContextOptionsDescriptor.ServiceType,
+                oldDbContextOptionsDescriptor.ServiceKey,
+                factory: (sp, key) =>
+                {
+                    var dbContextOptions = oldDbContextOptionsDescriptor.ImplementationFactory?.Invoke(sp) as DbContextOptions<TContext>;
+
+                    var optionsBuilder = dbContextOptions != null
+                        ? new DbContextOptionsBuilder<TContext>(dbContextOptions)
+                        : new DbContextOptionsBuilder<TContext>();
+
+                    optionsBuilder.UseSqlServer(options => options.EnableRetryOnFailure());
+
+                    return optionsBuilder.Options;
+                },
+                oldDbContextOptionsDescriptor.Lifetime
+                );
+
+            builder.Services.Add(dbContextOptionsDescriptor);
```

Comment:
Let's extract this into a common method so we don't have to duplicate this logic across all 5 EF components.